File: web/src/mocks/handlers/user-activity.ts
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import { activityStatus } from "@/features/core/users/activity/lib";
import { http, HttpResponse } from "msw";

// Fixed UTC fixture clock; demo data stays deterministic across machines and dates.
const NOW = "2026-09-16T12:00:00Z";
const dateAt = (days: number) => new Date(Date.parse(NOW) - days * 86_400_000).toISOString();
export const activityUsers = Array.from({ length: 225 }, (_, index) => {
  const days = [0, 1, 6, 7, 29, 30, 89, 120, null][index % 9] ?? null;
  const events =
    days === null
      ? []
      : Array.from({ length: 12 }, (_, offset) => {
          // Vary earlier visits across weekdays while preserving each user's last sign-in.
          const eventDays = days + offset * 7 - (offset === 0 ? 0 : (index + offset) % 5);
          const weekday = new Date(dateAt(eventDays)).getUTCDay();
          return {
            days: eventDays,
            count: weekday === 0 || weekday === 6 ? 1 : 1 + (index % 3),
          };
        });
  return {
    user_id: `activity-${index}`,
    name: `Activity User ${String(index + 1).padStart(3, "0")}`,
    email: `activity${index + 1}@example.org`,
    role_names: [["Staff"], ["Student"], ["Researcher"]][index % 3],
    created_at: dateAt(100 + index),
    last_login: days === null ? null : dateAt(days),
    inactive_days: days,
    login_count: events.reduce((sum, event) => sum + event.count, 0),
    login_day_count: events.length,
    current_streak: days !== null && days <= 1 ? 1 : 0,
    events,
  };
});
export function activityAnalytics(windowDays: number, userID?: string) {
  const users = userID ? activityUsers.filter((user) => user.user_id === userID) : activityUsers;
  const trend = Array.from({ length: windowDays }, (_, offset) => {
    const days = windowDays - offset - 1;
    return {
      date: dateAt(days).slice(0, 10),
      active_users: users.filter((user) => user.events.some((event) => event.days === days)).length,
      login_count: users.reduce(
        (sum, user) => sum + (user.events.find((event) => event.days === days)?.count ?? 0),
        0,
      ),
    };
  });
  return {
    generated_at: NOW,
    prior_active_users: users.filter((user) =>
      user.events.some((event) => event.days >= windowDays && event.days < 2 * windowDays),
    ).length,
    dormant_over_90_days: users.filter(
      (user) =>
        user.inactive_days !== null && user.inactive_days >= windowDays && user.inactive_days > 90,
    ).length,
    oldest_never_created_at:
      users
        .filter((user) => user.last_login === null)
        .map((user) => user.created_at)
        .sort()[0] ?? null,
    window_days: windowDays,
    total_users: users.length,
    users_ever_logged_in: users.filter((user) => user.last_login !== null).length,
    active_users: users.filter(
      (user) => user.inactive_days !== null && user.inactive_days < windowDays,
    ).length,
    lifetime_login_count: users.reduce((sum, user) => sum + user.login_count, 0),
    lifetime_active_days: users.reduce((sum, user) => sum + user.login_day_count, 0),
    window_login_count: trend.reduce((sum, point) => sum + point.login_count, 0),
    window_active_days: trend.reduce((sum, point) => sum + point.active_users, 0),
    trend,
  };
}
export function activityList(url: URL) {
  const windowDays = Number(url.searchParams.get("window") ?? 30);
  const status = url.searchParams.get("status") ?? "all";
  const query = (url.searchParams.get("query") ?? "").toLowerCase();
  const sort = url.searchParams.get("sort") ?? "last_login";
  const direction = url.searchParams.get("direction") === "asc" ? 1 : -1;
  const limit = Number(url.searchParams.get("limit") ?? 10);
  const offset = Number(url.searchParams.get("offset") ?? 0);
  const items = activityUsers
    .map(({ events, ...user }) => ({
      ...user,
      window_login_count: events
        .filter((event) => event.days < windowDays)
        .reduce((sum, event) => sum + event.count, 0),
    }))
    .filter(
      (user) =>
        (status === "all" || activityStatus(user, windowDays) === status) &&
        `${user.name} ${user.email}`.toLowerCase().includes(query),
    )
    .sort((a, b) => {
      const key = sort === "name" || sort === "login_count" ? sort : "last_login";
      const left = a[key];
      const right = b[key];
      if (left === right) return a.user_id.localeCompare(b.user_id);
      if (left === null) return 1;
      if (right === null) return -1;
      return (left < right ? -1 : 1) * direction;
    });
  return {
    items: items.slice(offset, offset + limit),
    total: items.length,
    limit,
    offset,
    window_days: windowDays,
    status,
  };
}
export const userActivityHandlers = [
  http.get("*/api/v1/users/activity", ({ request }) =>
    HttpResponse.json(activityList(new URL(request.url))),
  ),
  http.get("*/api/v1/users/activity/analytics", ({ request }) =>
    HttpResponse.json(
      activityAnalytics(Number(new URL(request.url).searchParams.get("window") ?? 30)),
    ),
  ),
  http.get("*/api/v1/users/:id/activity/analytics", ({ request, params }) => {
    if (!activityUsers.some((user) => user.user_id === params.id))
      return HttpResponse.json({ error: "Not found" }, { status: 404 });
    return HttpResponse.json(
      activityAnalytics(
        Number(new URL(request.url).searchParams.get("window") ?? 30),
        String(params.id),
      ),
    );
  }),
];

File: internal/server/user_activity.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package server

import (
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"strconv"
	"time"

	"github.com/apache/airavata-custos/internal/store"
	"github.com/apache/airavata-custos/pkg/common"
	"github.com/apache/airavata-custos/pkg/identity"
	"github.com/apache/airavata-custos/pkg/service"
)

// UserActivityListResponse is the activity audit envelope. It echoes the
// normalized window and status so the portal can reject a response from a
// backend that ignored those filters.
type UserActivityListResponse struct {
	Items      []store.UserActivityRow `json:"items"`
	Total      int                     `json:"total"`
	Limit      int                     `json:"limit"`
	Offset     int                     `json:"offset"`
	WindowDays int                     `json:"window_days"`
	Status     string                  `json:"status"`
}

// @Summary	List sign-in activity for users with an OIDC-linked identity
// @Description	One row per user with an OIDC-linked identity. `window` and `status` are applied before pagination, and the response echoes both so a caller can detect an unfiltered answer. Sign-in statuses are disjoint: active has signed in within the window, dormant has signed in earlier, never has no recorded sign-in.
// @Tags	Users
// @Security	BearerAuth
// @Produce	json
// @Param	window	query	integer	true	"Activity window in user-local calendar days (1-365)"
// @Param	status	query	string	true	"all | active | dormant | never"
// @Param	query	query	string	false	"Substring match on name or email"
// @Param	limit	query	integer	false	"Page size (default 10, max 200)"
// @Param	offset	query	integer	false	"Page offset (default 0)"
// @Param	sort	query	string	false	"name | last_login | login_count (default last_login)"
// @Param	direction	query	string	false	"asc | desc (default desc)"
// @Success	200	{object}	UserActivityListResponse
// @Failure	400	{object}	object{error=string}	"Invalid window, status, sort or direction"
// @Failure	403	{object}	object{code=string,message=string}	"Caller lacks core:users:activity:read"
// @Router	/users/activity [get]
func (s *Server) listUserActivity(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query()
	window, err := activityWindowParam(q.Get("window"), 0)
	if err != nil {
		common.WriteError(w, http.StatusBadRequest, err)
		return
	}
	page, err := s.svc.ListUserActivity(r.Context(), store.UserActivityFilter{
		WindowDays: window,
		Status:     q.Get("status"),
		Query:      q.Get("query"),
		Limit:      atoiOr(q.Get("limit"), 0),
		Offset:     atoiOr(q.Get("offset"), 0),
		Sort:       q.Get("sort"),
		Direction:  q.Get("direction"),
	})
	if err != nil {
		common.WriteServiceError(w, err)
		return
	}
	common.WriteJSON(w, http.StatusOK, UserActivityListResponse{
		Items:      page.Items,
		Total:      page.Total,
		Limit:      page.Limit,
		Offset:     page.Offset,
		WindowDays: page.WindowDays,
		Status:     page.Status,
	})
}

// @Summary	Aggregate sign-in analytics for all users with an OIDC-linked identity
// @Description	Population-wide sign-in activity over the requested window. Supplemental fields (`prior_active_users`, `dormant_over_90_days`, `oldest_never_created_at`) may be null when the population cannot produce them.
// @Tags	Users
// @Security	BearerAuth
// @Produce	json
// @Param	window	query	integer	false	"Activity window in user-local calendar days (1-365, default 30)"
// @Success	200	{object}	store.UserActivityAnalytics
// @Failure	400	{object}	object{error=string}	"Invalid window"
// @Failure	403	{object}	object{code=string,message=string}	"Caller lacks core:users:activity:read"
// @Router	/users/activity/analytics [get]
func (s *Server) getActivityAnalytics(w http.ResponseWriter, r *http.Request) {
	window, err := activityWindowParam(r.URL.Query().Get("window"), defaultActivityWindow)
	if err != nil {
		common.WriteError(w, http.StatusBadRequest, err)
		return
	}
	analytics, err := s.svc.GetActivityAnalytics(r.Context(), window)
	if err != nil {
		common.WriteServiceError(w, err)
		return
	}
	common.WriteJSON(w, http.StatusOK, analytics)
}

// @Summary	Aggregate sign-in analytics for one user with an OIDC-linked identity
// @Tags	Users
// @Security	BearerAuth
// @Produce	json
// @Param	id	path	string	true	"User ID"
// @Param	window	query	integer	false	"Activity window in user-local calendar days (1-365, default 30)"
// @Success	200	{object}	store.UserActivityAnalytics
// @Failure	400	{object}	object{error=string}	"Invalid window"
// @Failure	403	{object}	object{code=string,message=string}	"Caller lacks core:users:activity:read"
// @Failure	404	{object}	object{error=string}	"No such OIDC-linked user"
// @Router	/users/{id}/activity/analytics [get]
func (s *Server) getSelectedUserActivityAnalytics(w http.ResponseWriter, r *http.Request) {
	window, err := activityWindowParam(r.URL.Query().Get("window"), defaultActivityWindow)
	if err != nil {
		common.WriteError(w, http.StatusBadRequest, err)
		return
	}
	analytics, err := s.svc.GetUserActivityAnalytics(r.Context(), r.PathValue("id"), window)
	if err != nil {
		common.WriteServiceError(w, err)
		return
	}
	common.WriteJSON(w, http.StatusOK, analytics)
}

const defaultActivityWindow = 30

// activityWindowParam rejects a malformed window instead of falling back to a
// default, so the portal never renders one window's data under another's label.
func activityWindowParam(raw string, fallback int) (int, error) {
	if raw == "" {
		if fallback == 0 {
			return 0, errors.New("window is required")
		}
		return fallback, nil
	}
	window, err := strconv.Atoi(raw)
	if err != nil {
		return 0, fmt.Errorf("window must be a whole number of days: %w", err)
	}
	return window, nil
}

// recordLoginEventRequest is intentionally empty: the portal posts `{}` and
// the server derives every recorded fact from the verified bearer token.
type recordLoginEventRequest struct{}

// @Summary	Record the caller's sign-in
// @Description	Records one sign-in for the authenticated caller. Identity and timing come from the verified OIDC token, never the request body, and the body must be empty or `{}`. Repeating the call for the same token is idempotent and answers 200 without changing any counts.
// @Tags	Users
// @Security	BearerAuth
// @Accept	json
// @Produce	json
// @Success	201	{object}	service.RecordLoginEventResult	"Sign-in recorded"
// @Success	200	{object}	service.RecordLoginEventResult	"Already recorded for this token"
// @Failure	400	{object}	object{error=string}	"Body is not empty, or the token carries no usable session evidence"
// @Failure	401	{object}	object{error=string}	"No authenticated caller"
// @Router	/me/login-events [post]
func (s *Server) recordLoginEvent(w http.ResponseWriter, r *http.Request) {
	var req recordLoginEventRequest
	if err := common.DecodeJSON(r, &req); err != nil && !errors.Is(err, io.EOF) {
		common.WriteError(w, http.StatusBadRequest, err)
		return
	}
	caller := requireCaller(w, r)
	if caller == nil {
		return
	}
	input, err := trustedLoginEvent(caller)
	if err != nil {
		common.WriteError(w, http.StatusBadRequest, err)
		return
	}
	result, err := s.svc.RecordLoginEvent(r.Context(), caller.UserID, input)
	if err != nil {
		common.WriteServiceError(w, err)
		return
	}
	status := http.StatusOK
	if result.Recorded {
		status = http.StatusCreated
	} else {
		slog.DebugContext(r.Context(), "login event already recorded", "user_id", caller.UserID)
	}
	common.WriteJSON(w, status, result)
}

// trustedLoginEvent derives the dedup key and timestamp from verified claims.
// A token identifier plus its issue time is preferred: two portal sign-ins
// that share one identity-provider session still produce distinct tokens,
// while the session claims alone would collapse them into one event.
func trustedLoginEvent(caller *identity.Caller) (service.RecordLoginEventInput, error) {
	if caller.TokenID != "" && caller.IssuedAt > 0 {
		return service.RecordLoginEventInput{
			EventKey:   hashedEventKey("token", caller.Issuer, caller.Subject, caller.TokenID, caller.IssuedAt),
			OccurredAt: time.Unix(caller.IssuedAt, 0).UTC(),
			Provider:   "oidc",
			SessionID:  caller.SessionID,
		}, nil
	}
	if caller.SessionID != "" && caller.AuthTime > 0 {
		return service.RecordLoginEventInput{
			EventKey:   hashedEventKey("session", caller.Issuer, caller.Subject, caller.SessionID, caller.AuthTime),
			OccurredAt: time.Unix(caller.AuthTime, 0).UTC(),
			Provider:   "oidc",
			SessionID:  caller.SessionID,
		}, nil
	}
	return service.RecordLoginEventInput{}, errors.New(
		"token carries no session evidence: jti with iat, or sid with auth_time, is required")
}

// hashedEventKey builds a collision-resistant key over NUL-separated claim
// values so a value containing the separator cannot forge another key.
func hashedEventKey(kind, issuer, subject, sessionID string, timestamp int64) string {
	canonical := fmt.Sprintf("v1\x00%s\x00%s\x00%s\x00%s\x00%d", kind, issuer, subject, sessionID, timestamp)
	sum := sha256.Sum256([]byte(canonical))
	return "v1:" + hex.EncodeToString(sum[:])
}
